src/orion/core/evc/tree.py

```diff
@@ -0,0 +1,389 @@
+
+
+class TreeNode(object):
+    r"""Tree node data structure
+
+    Nodes have an attribute item to carry arbitrary information. A node may only have one parent and
+    can have as many children as desired.
+
+    Parents can be set at initialiation or via `node.set_parent`. Setting a parent automatically set
+    the current node as the child of the parent.
+
+    Children can be set at initialiation or via `node.add_children`. Setting children automatically
+    set their parent as the current node.
+
+    Tree of nodes are iterable, by default with preorder traversal.
+
+    .. seealso::
+        `orion.core.evc.tree.PreOrderTraversal`
+        `orion.core.evc.tree.DepthFirstTraversal`
+
+    Attributes
+    ----------
+    item: object
+        Can be anything
+    parent: None or instance of `orion.core.evc.tree.TreeNode`
+        The parent of the current node, None if the current node is the root.
+    children: None or list of instances of `orion.core.evc.tree.TreeNode`
+        The children of the curent node.
+    root: instance of `orion.core.evc.tree.TreeNode`
+        The top node of the current tree. The root node returns itself.
+
+    Example
+    -------
+    .. code-block:: python
+        :linenos:
+
+        a = TreeNode("a")
+        b = TreeNode("b", a)
+        c = TreeNode("c", a)
+        d = TreeNode("d", a)
+        e = TreeNode("e", a)
+
+        f = TreeNode("f", b)
+        g = TreeNode("g", b)
+
+        e = TreeNode("h", e)
+
+        # Gives this tree
+        # a
+        # |   \  \   \
+        # b    c  d   e
+        # | \         |
+        # f  g        h
+
+        g.set_parent(e)
+
+        # Gives this tree
+        # a
+        # |  \  \   \
+        # b   c  d   e
+        # |          | \
+        # f          h g
+
+        c.add_children(h, g)
+
+        # Gives this tree
+        # a
+        # |  \    \   \
+        # b   c    d   e
+        # |   | \
+        # f   h  g
+
+        a.drop_children(c)
+
+        # Gives this tree
+        # a
+        # |  \   \
+        # b   d   e
+        # |
+        # f
+
+    """
+
+    __slots__ = ('_item', '_parent', '_children')
+
+    def __init__(self, item, parent=None, children=tuple()):
+        """Initialize node with item, parent and children
+
+        .. seealso::
+            :class:`orion.core.evc.tree.TreeNode` for information about the attributes
+        """
+        self._item = item
+        self._parent = None
+        self._children = []
+
+        self.set_parent(parent)
+        self.add_children(*children)
+
+    @property
+    def item(self):
+        """Get item of the node which may contain arbitrary objects"""
+        return self._item
+
+    @item.setter
+    def item(self, new_item):
+        """Set item of the node with arbitrary objects"""
+        self._item = new_item
+
+    @property
+    def parent(self):
+        """Get parent of the node, None if no parent"""
+        return self._parent
+
+    def drop_parent(self):
```

Comment:
Yeah if you want to avoid infinit recursion you need to either reimplement the whole A,B,C shuffling in both parent and children side, or you implement it only on either parent or children side and call from the other way around. I decided to do the A,B,C shuffling the children and call children methods from parent methods.